src/Testing/CodeHacking/README.md

```diff
@@ -0,0 +1,177 @@
+ * Hack classes inside the `Hack` folders: some predefined frequently used hacks.
+
+## How to use
+
+Let's go through an example.
+
+First, create a file named `class-wc-admin-hello-worlder` in `includes/admin/class-wc-admin-hello-worlder.php` with the following code:
+
+```
+<?php
+
+class WC_Admin_Hello_Worlder {
+
+	public function say_hello( $to_who ) {
+		Logger::log_hello( $to_who );
+		$site_name = get_option( 'site_name' );
+		return "Hello {$to_who}, welcome to {$site_name}!";
+	}
+}
+```
+
+This class has two bits that are difficult to unit test: the call to `Logger::log_hello` and the `get_option` invocation. Let's see how the code hacker can help us with that. 
+
+Create a file named `class-wc-tests-admin-hello-worlder.php` in `tests/unit-tests/admin` with this code:
+
+```
+<?php
+
+use Automattic\WooCommerce\Testing\CodeHacking\CodeHacker;
+use Automattic\WooCommerce\Testing\CodeHacking\Hacks\StaticMockerHack;
+use Automattic\WooCommerce\Testing\CodeHacking\Hacks\FunctionsMockerHack;
+
+class LoggerMock {
+
+	public static $_logged = null;
+
+	public static function log_hello( $to_who ) {
+		self::$_logged = $to_who;
+	}
+}
+
+class FunctionsMock {
+
+	public static $_option_requested = null;
+	public static $_option_value     = null;
+
+	public static function get_option( $option, $default = false ) {
+		self::$_option_requested = $option;
+		return self::$_option_value;
+	}
+}
+
+class WC_Tests_Admin_Hello_Worlder extends WC_Unit_Test_Case {
+
+	public static function before_test_say_hello() {
+		CodeHacker::add_hack( new StaticMockerHack( 'Logger', 'LoggerMock' ) );
+		CodeHacker::add_hack( new FunctionsMockerHack( 'FunctionsMock' ) );
+		CodeHacker::enable();
+	}
+
+	public function test_say_hello() {
+		FunctionsMock::$_option_value = 'MSX world';
+
+		$sut    = new WC_Admin_Hello_Worlder();
+		$actual = $sut->say_hello( 'Nestor' );
+
+		$this->assertEquals( 'Hello Nestor, welcome to MSX world!', $actual );
+		$this->assertEquals( 'site_name', FunctionsMock::$_option_requested );
+		$this->assertEquals( 'Nestor', LoggerMock::$_logged );
+	}
+}
+```
+
+Then run `vendor/bin/phpunit tests/unit-tests/admin/class-wc-tests-admin-hello-worlder.php` and see the magic happen. Note that this works because `CodeHackerTestHook` has been registered in `phpunit.xml`.
+
+As you can see, the basic mechanism consists of creating a `public static before_[test_method_name]` method in the test class, where you register whatever hacks you need with `CodeHacker::add_hack` and finally you invoke `CodeHacker::enable()` to enable the hacking. `StaticMockerHack` and `FunctionsMockerHack` are two of the predefined hack classes inside the `Hack` folder, see their source code for details on how they work and how to use them. 
+
+You can define a `before_all` method too, which will run before all the tests (and before the `before_[test_method_name]` method).
```

Comment:
let's call it something other than `before_all` if using setup is not possible,  because these two naming conventions are used interchangeably across languages.

Reply:
Ok, what would you suggest? Maybe `before_any_test`?

Reply:
let's call them: `add_hack_for_[test_method_name]` and `add_hack_for_all`. I want to make them specific because for all normal use cases, `setup` and `teardown` should be used. wdyt?